File: API/0051_Volume_Weighted_Price_Breakout/VolumeWeightedPriceBreakoutStrategy.cs
using System;
using System.Collections.Generic;

using StockSharp.Algo.Indicators;
using StockSharp.Algo.Strategies;
using StockSharp.BusinessEntities;
using StockSharp.Messages;

namespace StockSharp.Samples.Strategies
{
    /// <summary>
    /// Volume Weighted Price Breakout Strategy
    /// Long entry: Price rises above the volume-weighted average price over N periods
    /// Short entry: Price falls below the volume-weighted average price over N periods
    /// Exit: Price crosses MA in the opposite direction
    /// </summary>
    public class VolumeWeightedPriceBreakoutStrategy : Strategy
    {
        private readonly StrategyParam<int> _maPeriod;
        private readonly StrategyParam<int> _vwapPeriod;
        private readonly StrategyParam<DataType> _candleType;

        /// <summary>
        /// MA Period
        /// </summary>
        public int MAPeriod
        {
            get => _maPeriod.Value;
            set => _maPeriod.Value = value;
        }

        /// <summary>
        /// VWAP Period
        /// </summary>
        public int VWAPPeriod
        {
            get => _vwapPeriod.Value;
            set => _vwapPeriod.Value = value;
        }

        /// <summary>
        /// Candle type for strategy calculation
        /// </summary>
        public DataType CandleType
        {
            get => _candleType.Value;
            set => _candleType.Value = value;
        }

        /// <summary>
        /// Initialize <see cref="VolumeWeightedPriceBreakoutStrategy"/>.
        /// </summary>
        public VolumeWeightedPriceBreakoutStrategy()
        {
            _maPeriod = Param(nameof(MAPeriod), 20)
                .SetGreaterThanZero()
                .SetDisplay("MA Period", "Period for Moving Average calculation", "Strategy Parameters")
                .SetCanOptimize(true)
                .SetOptimize(10, 50, 10);

            _vwapPeriod = Param(nameof(VWAPPeriod), 20)
                .SetGreaterThanZero()
                .SetDisplay("VWAP Period", "Period for Volume Weighted Average Price calculation", "Strategy Parameters")
                .SetCanOptimize(true)
                .SetOptimize(10, 30, 5);

            _candleType = Param(nameof(CandleType), TimeSpan.FromMinutes(5).TimeFrame())
                .SetDisplay("Candle Type", "Type of candles for strategy calculation", "Strategy Parameters");
        }

        /// <inheritdoc />
        public override IEnumerable<(Security sec, DataType dt)> GetWorkingSecurities()
        {
            return [(Security, CandleType)];
        }

        /// <inheritdoc />
        protected override void OnStarted(DateTimeOffset time)
        {
            base.OnStarted(time);

            // Create indicators
            var ma = new SimpleMovingAverage { Length = MAPeriod };
            var vwma = new VolumeWeightedMovingAverage { Length = VWAPPeriod };

            // Create subscription and bind indicators
            var subscription = SubscribeCandles(CandleType);
            
            subscription
                .Bind(ma, vwma, ProcessCandle)
                .Start();

            // Configure protection
            StartProtection(
                takeProfit: new Unit(3, UnitTypes.Percent),
                stopLoss: new Unit(2, UnitTypes.Percent)
            );

            // Setup chart visualization
            var area = CreateChartArea();
            if (area != null)
            {
                DrawCandles(area, subscription);
                DrawIndicator(area, ma);
                DrawIndicator(area, vwma);
                DrawOwnTrades(area);
            }
        }

        private void ProcessCandle(ICandleMessage candle, decimal maValue, decimal vwmaValue)
        {
            // Skip unfinished candles
            if (candle.State != CandleStates.Finished)
                return;

            // Check if strategy is ready to trade
            if (!IsFormedAndOnlineAndAllowTrading())
                return;

            // Log current values
            LogInfo($"Candle Close: {candle.ClosePrice}, MA: {maValue}, VWMA: {vwmaValue}");

            // Trading logic:
            // Long: Price above VWMA
            if (candle.ClosePrice > vwmaValue && Position <= 0)
            {
                LogInfo($"Buy Signal: Price ({candle.ClosePrice}) > VWMA ({vwmaValue})");
                BuyMarket(Volume + Math.Abs(Position));
            }
            // Short: Price below VWMA
            else if (candle.ClosePrice < vwmaValue && Position >= 0)
            {
                LogInfo($"Sell Signal: Price ({candle.ClosePrice}) < VWMA ({vwmaValue})");
                SellMarket(Volume + Math.Abs(Position));
            }
            
            // Exit logic: Price crosses MA in the opposite direction
            if (Position > 0 && candle.ClosePrice < maValue)
            {
                LogInfo($"Exit Long: Price ({candle.ClosePrice}) < MA ({maValue})");
                SellMarket(Math.Abs(Position));
            }
            else if (Position < 0 && candle.ClosePrice > maValue)
            {
                LogInfo($"Exit Short: Price ({candle.ClosePrice}) > MA ({maValue})");
                BuyMarket(Math.Abs(Position));
            }
        }
    }
}

File: API/0051_Volume_Weighted_Price_Breakout/PY/volume_weighted_price_breakout_strategy.py
import clr

clr.AddReference("StockSharp.Messages")
clr.AddReference("StockSharp.Algo")

from System import TimeSpan
from StockSharp.Messages import DataType, CandleStates
from StockSharp.Algo.Indicators import SimpleMovingAverage, VolumeWeightedMovingAverage
from StockSharp.Algo.Strategies import Strategy

class volume_weighted_price_breakout_strategy(Strategy):
    """
    Volume Weighted Price Breakout Strategy.
    Long entry: Price rises above VWMA.
    Short entry: Price falls below VWMA.
    Exit: Price crosses MA in the opposite direction.
    """

    def __init__(self):
        super(volume_weighted_price_breakout_strategy, self).__init__()
        self._ma_period = self.Param("MAPeriod", 20).SetDisplay("MA Period", "Period for Moving Average", "Indicators")
        self._vwap_period = self.Param("VWAPPeriod", 20).SetDisplay("VWAP Period", "Period for VWMA", "Indicators")
        self._candle_type = self.Param("CandleType", DataType.TimeFrame(TimeSpan.FromMinutes(1))).SetDisplay("Candle Type", "Type of candles to use", "General")
        self._cooldown_bars = self.Param("CooldownBars", 500).SetDisplay("Cooldown Bars", "Bars to wait between trades", "General")

        self._cooldown = 0

    @property
    def candle_type(self):
        return self._candle_type.Value

    def OnReseted(self):
        super(volume_weighted_price_breakout_strategy, self).OnReseted()
        self._cooldown = 0

    def OnStarted(self, time):
        super(volume_weighted_price_breakout_strategy, self).OnStarted(time)

        self._cooldown = 0

        ma = SimpleMovingAverage()
        ma.Length = self._ma_period.Value
        vwma = VolumeWeightedMovingAverage()
        vwma.Length = self._vwap_period.Value

        subscription = self.SubscribeCandles(self.candle_type)
        subscription.Bind(ma, vwma, self._process_candle).Start()

        area = self.CreateChartArea()
        if area is not None:
            self.DrawCandles(area, subscription)
            self.DrawIndicator(area, ma)
            self.DrawIndicator(area, vwma)
            self.DrawOwnTrades(area)

    def _process_candle(self, candle, ma_val, vwma_val):
        if candle.State != CandleStates.Finished:
            return

        if self._cooldown > 0:
            self._cooldown -= 1
            return

        close = float(candle.ClosePrice)
        mv = float(ma_val)
        vv = float(vwma_val)
        cd = self._cooldown_bars.Value

        if self.Position == 0:
            if close > vv:
                self.BuyMarket()
                self._cooldown = cd
            elif close < vv:
                self.SellMarket()
                self._cooldown = cd
        elif self.Position > 0 and close < mv:
            self.SellMarket()
            self._cooldown = cd
        elif self.Position < 0 and close > mv:
            self.BuyMarket()
            self._cooldown = cd

    def CreateClone(self):
        return volume_weighted_price_breakout_strategy()
